cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
   return trajectory_nodes_list;
 }
 
+visualization_msgs::MarkerArray MapBuilderBridge::GetConstraintsList() {
+  visualization_msgs::MarkerArray constraints_list;
+  int marker_id = 0;
+  const ros::Time now = ros::Time::now();
+  visualization_msgs::Marker constraint_intra_marker;
+  constraint_intra_marker.id = marker_id++;
+  constraint_intra_marker.ns = "Intra constraints";
+  constraint_intra_marker.type = visualization_msgs::Marker::LINE_LIST;
+  constraint_intra_marker.header.stamp = now;
+  constraint_intra_marker.header.frame_id = node_options_.map_frame;
+  constraint_intra_marker.scale.x = kConstraintLineListMarkerScale;
+  constraint_intra_marker.pose.orientation.w = 1.0;
+
+  visualization_msgs::Marker residual_intra_marker = constraint_intra_marker;
+  residual_intra_marker.id = marker_id++;
+  residual_intra_marker.ns = "Intra residuals";
+  residual_intra_marker.pose.position.z = 0.1;
+
+  visualization_msgs::Marker constraint_inter_marker = constraint_intra_marker;
+  constraint_inter_marker.id = marker_id++;
+  constraint_inter_marker.ns = "Inter constraints";
+  constraint_inter_marker.pose.position.z = 0.1;
+
+  visualization_msgs::Marker residual_inter_marker = constraint_intra_marker;
+  residual_inter_marker.id = marker_id++;
+  residual_inter_marker.ns = "Inter residuals";
+  residual_inter_marker.pose.position.z = 0.1;
+
+  const auto all_trajectory_nodes =
+      map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+  const auto all_submap_data =
+      map_builder_.sparse_pose_graph()->GetAllSubmapData();
+  const auto constraints = map_builder_.sparse_pose_graph()->constraints();
+
+  // Color mapping for submaps of various trajectories - add trajectory id
+  // to ensure different starting colors.
+  const auto submap_color_id_map =
+      [](const cartographer::mapping::SubmapId submap_id) -> int {
+    return submap_id.submap_index + submap_id.trajectory_id + 20;
+  };
+
+  for (const auto& constraint : constraints) {
+    visualization_msgs::Marker *constraint_marker, *residual_marker;
+    std_msgs::ColorRGBA color_constraint, color_residual;
+    if (constraint.tag ==
+        cartographer::mapping::SparsePoseGraph::Constraint::INTRA_SUBMAP) {
+      constraint_marker = &constraint_intra_marker;
+      residual_marker = &residual_intra_marker;
+      color_constraint = GetColor(submap_color_id_map(constraint.submap_id));
+      color_residual.a = 1.0;
```

Comment:
How would you differentiate visually the different types of constraints and markers? Since the intra constraints are the most numerous, I decided to use the color scheme for their markers; for other types of markers, I used fixed colors (e.g. so you know that loop closure constraints are bright yellow).